spec.bs

```diff
@@ -750,19 +769,45 @@ A <dfn export for=fencedframetype>fenced frame reporter</dfn> is a [=struct=] wi
      </dl>
 </dl>
 
+A <dfn export for=fencedframetype>destination enum event</dfn> is a [=struct=] with the following
+[=struct/items=]:
+<dl export dfn-for="destination enum event">
+  : <dfn>type</dfn>
+  :: a [=string=]
+
+  : <dfn>data</dfn>
+  :: a [=string=]
+</dl>
+
+A <dfn export for=fencedframetype>destination URL event</dfn> is a [=URL=].
+
+A <dfn export for=fencedframetype>destination event</dfn> is either a
+[=fencedframetype/destination enum event=] or a [=fencedframetype/destination URL event=].
+
 <div algorithm>
-  In order to <dfn>send a beacon</dfn> with a [=map=] |destination map| whose [=map/keys=] are
-  [=strings=] and whose [=map/values=] are [=urls=], a [=string=] |eventType|, and a [=string=]
-  |eventData|, run these steps:
+  In order to <dfn>send a beacon</dfn> with a [=fencedframetype/reporting destination info=]
+  |destination info| and a [=fencedframetype/destination event=] |event|, run these steps:
 
-  1. If |destination map|[|eventType|] does not [=map/exist=], return.
+  1. Let |destination url| be an empty [=string=].
+  1. If |event| is a [=fencedframetype/destination enum event=]:
+     1. Let |destination map| be |destination info|'s
+        [=reporting destination info/reporting url map=].
+     1. |destination map|[|eventType|] does not [=map/exist=], return.
 
-  1. Let |destination url| be |destination map|[|eventType|].
+     1. Set |destination url| to |destination map|[|eventType|].
+
+  1. Otherwise:
+     1. Assert that |event| is a [=fencedframetype/destination URL event=].
+     1. Let |macro map| be |destination info|'s [=reporting destination info/reporting macro map=].
+     1. If |macro map| is null, return.
+     1. Set |destination url| to |event|.
+     1. TODO: Substitute macros from |macro map| into |destination url|.
```

Comment:
@domfarolino Do you know of any existing spec for functionality like this? (take a map of string -> string, and substitute the keys -> values in another string without recursive substitution)

Reply:
Hmm, the only kind of substitution like this that comes to mind is when we substitute format specifiers in e.g., `console.log('My number is: %f', .435)`, which is in https://console.spec.whatwg.org/#formatter. That's a pretty manual process, and I don't know of any preexisting reusable utility for this unfortunately, so you'll probably just have to do it that way.

Reply:
Factored it out into a TODO. The exact semantics of this are not super important right now.